x-pack/legacy/plugins/security/server/lib/authentication/authentication_result.ts

```diff
@@ -62,42 +62,58 @@ export class AuthenticationResult {
   /**
    * Produces `AuthenticationResult` for the case when authentication succeeds.
    * @param user User information retrieved as a result of successful authentication attempt.
+   * @param [authResponseHeaders] Optional dictionary of the HTTP headers with authentication
+   * information that should be specified in the response we send to the client request.
    * @param [state] Optional state to be stored and reused for the next request.
    */
-  public static succeeded(user: AuthenticatedUser, state?: unknown) {
+  public static succeeded(
+    user: AuthenticatedUser,
+    {
+      authResponseHeaders,
+      state,
+    }: Pick<AuthenticationOptions, 'authResponseHeaders' | 'state'> = {}
+  ) {
     if (!user) {
       throw new Error('User should be specified.');
     }
 
-    return new AuthenticationResult(AuthenticationResultStatus.Succeeded, { user, state });
+    return new AuthenticationResult(AuthenticationResultStatus.Succeeded, {
+      user,
+      authResponseHeaders,
+      state,
+    });
   }
 
   /**
    * Produces `AuthenticationResult` for the case when authentication fails.
    * @param error Error that occurred during authentication attempt.
-   * @param [challenges] Optional list of the challenges that will be returned to the user within
-   * `WWW-Authenticate` HTTP header. Multiple challenges will result in multiple headers (one per
-   * challenge) as it's better supported by the browsers than comma separated list within a single
-   * header. Challenges can only be set for errors with `401` error status.
-   */
-  public static failed(error: Error, challenges?: string[]) {
+   * @param [authResponseHeaders] Optional dictionary of the HTTP headers with authentication related
+   * information (e.g. `WWW-Authenticate` with the challenges) that should be specified in the
+   * response we send to the client request.
+   */
+  public static failed(
+    error: Error,
+    { authResponseHeaders }: Pick<AuthenticationOptions, 'authResponseHeaders'> = {}
+  ) {
     if (!error) {
       throw new Error('Error should be specified.');
     }
 
-    if (challenges != null && getErrorStatusCode(error) !== 401) {
-      throw new Error('Challenges can only be provided with `401 Unauthorized` errors.');
-    }
-
-    return new AuthenticationResult(AuthenticationResultStatus.Failed, { error, challenges });
+    return new AuthenticationResult(AuthenticationResultStatus.Failed, {
+      error,
+      authResponseHeaders,
+    });
   }
 
   /**
    * Produces `AuthenticationResult` for the case when authentication needs user to be redirected.
    * @param redirectURL URL that should be used to redirect user to complete authentication.
    * @param [state] Optional state to be stored and reused for the next request.
    */
-  public static redirectTo(redirectURL: string, state?: unknown) {
+  public static redirectTo(
+    redirectURL: string,
+    { state }: Pick<AuthenticationOptions, 'state'> = {}
```

Comment:
__note:__ just to have signature consistent with the rest of the methods in this class.